Check if address is valid

7h7 opened the pull request:
# Description

Check for value transactions that the last trit from the address is 0, otherwise it's invalid and the bundle isn't accepted by the node
https://github.com/iotaledger/iri/blob/1ec6dcc007a0cd119d6a9c9878842515390ced9b/src/main/java/com/iota/iri/TransactionValidator.java#L179-L181

Fixes #866

## Type of change

- Bug fix (a non-breaking change which fixes an issue)

# Checklist:

- [x] My code follows the style guidelines for this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added tests that prove my fix is effective or that my feature works

7h7 committed:
Add isLastTritZero

7h7 committed:
Update transfers.js

7h7 committed:
Update errors.js

7h7 committed:
Rajiv's fault

7h7 committed:
Update transfers.js

7h7 committed:
Update translation.json

7h7 committed:
Update transfers.js

7h7 committed:
Update utils.spec.js

7h7 committed:
Update utils.spec.js

rajivshah3 commented on src/shared/__tests__/libs/iota/utils.spec.js:
```diff
@@ -10,6 +10,18 @@ import {
+    describe('#isLastTritZero', () => {
+        describe('when the last trit is 0', () => {
+            it('should return true', () => {
+                expect(isLastTritZero('D'.repeat(81)).to.equal(true);
```
```suggestion
                expect(isLastTritZero('D'.repeat(81))).to.equal(true);
```

rajivshah3 commented on src/shared/__tests__/libs/iota/utils.spec.js:
```diff
@@ -10,6 +10,18 @@ import {
+        });
+        describe('when the last trit is not 0', () => {
+            it('should return false', () => {
+                expect(isLastTritZero('E'.repeat(81)).to.equal(false);
```
```suggestion
                expect(isLastTritZero('E'.repeat(81))).to.equal(false);
```

rajivshah3 commented on src/shared/actions/transfers.js:
```diff
@@ -433,7 +433,10 @@ export const makeTransaction = (seedStore, receiveAddress, value, message, accou
         // Progressbar step => (Validating receive address)
         dispatch(setNextStepAsActive());
-
+        //Check the last trit for validity
```
Small stylistic change
```suggestion
        // Check the last trit for validity
```

rajivshah3 commented on src/shared/libs/iota/utils.js:
```diff
@@ -191,6 +191,21 @@ export const isValidAddress = (address) => {
+ * @returns {boolean}
+ */
+export const isLastTritZero = (address) => {
+    if (!/[E-V]/.test(address.slice(80,81)) {
```
```suggestion
    if (!/[E-V]/.test(address.slice(80, 81))) {
```

rajivshah3 commented on src/shared/actions/transfers.js:
```diff
@@ -433,7 +433,10 @@ export const makeTransaction = (seedStore, receiveAddress, value, message, accou
         dispatch(setNextStepAsActive());
-
+        //Check the last trit for validity
+        if(isLastTritZero(address) === false){
```
Small stylistic change
```suggestion
        if (isLastTritZero(address) === false){
```

rajivshah3 replied:
`isLastTritZero` needs to be imported [here](https://github.com/7h7/trinity-wallet/blob/develop/src/shared/actions/transfers.js#L28) as well (it won't let me comment on that line):
```js
import {
    withRetriesOnDifferentNodes,
    fetchRemoteNodes,
    getRandomNodes,
    throwIfNodeNotHealthy,
    isLastTritZero,
} from '../libs/iota/utils';
```

rajivshah3 commented on src/shared/__tests__/libs/iota/utils.spec.js:
```diff
@@ -10,6 +10,18 @@ import {
 } from '../../../libs/iota/utils';
```
`isLastTritZero` needs to be imported
```suggestion
    isLastTritZero,
} from '../../../libs/iota/utils';
```

rajivshah3 reviewed (changes requested):
Thanks for the PR @7h7 ! The following changes should make the CI pass

rajivshah3 committed:
Update src/shared/__tests__/libs/iota/utils.spec.js

Co-Authored-By: 7h7 <tho5671@web.de>

rajivshah3 committed:
Update src/shared/libs/iota/utils.js

Co-Authored-By: 7h7 <tho5671@web.de>

rajivshah3 committed:
Update src/shared/__tests__/libs/iota/utils.spec.js

Co-Authored-By: 7h7 <tho5671@web.de>

rajivshah3 committed:
Update src/shared/__tests__/libs/iota/utils.spec.js

Co-Authored-By: 7h7 <tho5671@web.de>

rajivshah3 committed:
Update src/shared/actions/transfers.js

Co-Authored-By: 7h7 <tho5671@web.de>

rajivshah3 committed:
Update src/shared/actions/transfers.js

Co-Authored-By: 7h7 <tho5671@web.de>

rajivshah3 committed:
Shared: Add isLastTritZeroAsync to the promise chain

- Add new method isLastTritZeroAsync, an async version of isLastTritZero
- Add it to the promise chain in `makeTransaction`

rajivshah3 committed:
Shared: Fix failing tests

laumair commented:
Under which circumstances would Trinity generate an address with a non-zero trit? 

I think this validation should also be added to [isBundle](https://github.com/iotaledger/iota.js/blob/master/lib/utils/utils.js#L388) so that these bundles are never added to Trinity.

rajivshah3 commented:
> Under which circumstances would Trinity generate an address with a non-zero trit?

I don’t think these addresses can be generated with a lib. They are normally human-generated. An example @7h7 gave was 
https://thetangle.org/address/TROLLTROLLTROLLTROLLTROLLTROLLTROLLTROLLTROLLTROLLTROLLTROLLTROLLTROLLTROLLTROLLT. You can try to send a value transfer to this address, but IRI will reject it. The problem is, at that point the transaction is signed and the signature is exposed to the node. 

> I think this validation should also be added to isBundle so that these bundles are never added to Trinity.

I agree

laumair commented:
> I don’t think these addresses can be generated with a lib. They are normally human-generated. An example @7h7 gave was
https://thetangle.org/address/TROLLTROLLTROLLTROLLTROLLTROLLTROLLTROLLTROLLTROLLTROLLTROLLTROLLTROLLTROLLTROLLT. You can try to send a value transfer to this address, but IRI will reject it. The problem is, at that point the transaction is signed and the signature is exposed to the node.

Okay thanks for the explanation. This check should be added for recipient addresses.

7h7 committed:
Merge pull request #1 from rajivshah3/last-trit-fixes

isLastTritZero fixes

rajivshah3 committed:
Merge branch 'develop' into 7h7-develop

7h7 committed:
Merge pull request #2 from rajivshah3/868-conflicts

Merge upstream develop branch

rajivshah3 reviewed (approved):
Looks good to me! Thanks @7h7

laumair commented on src/shared/libs/iota/utils.js:
```diff
@@ -191,6 +191,36 @@ export const isValidAddress = (address) => {
+ *
+ * @returns {boolean}
+ */
+export const isLastTritZeroAsync = (address) => {
```
I think adding an async util here is unnecessary. 

I can see the motivation for its addition and [usage](https://github.com/iotaledger/trinity-wallet/pull/868/files#diff-8ea5b2a5a0c274aaec60a3bbffa271c0R436) but you could also do:

```
return Promise.resolve(isLastTritZero(...)).then(() => {}).catch(() => {})
```

laumair commented on src/shared/libs/iota/utils.js:
```diff
@@ -191,6 +191,36 @@ export const isValidAddress = (address) => {
+ * @returns {boolean}
+ */
+export const isLastTritZero = (address) => {
+    if (!/[E-V]/.test(address.slice(80, 81))) {
```
Just a suggestion but if you're not doing any validation on the parameter perhaps this can be simplified to just `export const isLastTritZero = (address) => !/[E-V]/.test(address.slice(80, 81))` i.e., no need for explicit returns.

7h7 committed:
Update utils.js

7h7 committed:
Update utils.js

7h7 committed:
Update transfers.js

rajivshah3 commented on src/shared/actions/transfers.js:
```diff
@@ -427,124 +428,132 @@ export const makeTransaction = (seedStore, receiveAddress, value, message, accou
-                // Only update the local store with updated account information after this transaction is successfully completed.
-                accountState = newState;
+    // Check the last trit for validity
+    if (!lastTritIsZero(address)) {
```
We want this to be part of the promise chain (since there's a possibility that we throw an error here, we want it to throw and then go to the `catch` block and show the error message to the user), so I think the best approach would be similar to what we had before (but incorporate @laumair 's suggestions):
```js
return Promise.resolve(isLastTritZero(address)).then((lastTritIsZero) => {
  if(!lastTritIsZero) {
    throw new Error(Errors.INVALID_LAST_TRIT);
  }
  
  return isAnyAddressSpent()([address])
  ...
```
The reason why we use these [promises](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Using_promises) is twofold: it makes error handling a bit easier, while also allowing us to use asynchronous functions without our code looking like [this](http://callbackhell.com/) :trollface: . So this is essentially what we're doing:

1. Call `isLastTritZero`, passing `address` as a parameter
2. Store whatever `isLastTritZero` returns in a variable called `lastTritIsZero` (in our case, this will be either true or false)
3. Is the value of `lastTritIsZero` ["falsy"](https://developer.mozilla.org/en-US/docs/Glossary/Falsy)? (so in our case, we're essentially checking if `lastTritIsZero === false`)
 a. i. If so, throw an error
 a. ii. When the error is thrown, we'll skip all the way to the `catch` block so we can figure out what to do with the error https://github.com/iotaledger/trinity-wallet/blob/252a1ed3de6f453beef52f6c510c376979990e3f/src/shared/actions/transfers.js#L726
  b. i. If not, then we'll continue by calling `isAnyAddressSpent`

7h7 committed:
Update transfers.js

7h7 committed:
Update transfers.js

rajivshah3 committed:
Shared: Fix indentation and parentheses/brackets

laumair reviewed (approved):
@7h7 @rajivshah3  💯
